clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -632,9 +668,48 @@ synchronized boolean hasOngoingTransaction() {
         return currentState == State.IN_TRANSACTION || isCompleting() || hasAbortableError();
     }
 
-    synchronized boolean canRetryOutOfOrderSequenceException(ProducerBatch batch) {
-        return hasProducerId(batch.producerId()) && !hasUnresolvedSequence(batch.topicPartition) &&
-                (batch.sequenceHasBeenReset() || !isNextSequence(batch.topicPartition, batch.baseSequence()));
+    synchronized boolean canRetry(ProduceResponse.PartitionResponse response, ProducerBatch batch) {
+        if (!hasProducerId(batch.producerId()))
+            return false;
+
+        Errors error = response.error;
+        if (error == Errors.OUT_OF_ORDER_SEQUENCE_NUMBER && !hasUnresolvedSequence(batch.topicPartition) &&
```

Comment:
Shouldn't Errors.UNKNOWN_PRODUCER_ID be included here as well ?
Considering the UnknownProducerIdException is a subclass of OutOfOrderSequenceException.

Reply:
Actually, no. This is the part of the code where we want to disambiguate between `UNKNOWN_PRODUCER_ID` and `OUT_OF_ORDER_SEQUENCE_EXCEPTION`.